### PR TITLE
LightGBM - GBDT model earlystop bug fix

### PR DESCRIPTION
**Pull request by aaronrockmenezes**
## Description

Fix for #3559 

Added callbacks for early stopping for newer versions of LightGBM and XGBoost. 

## Type of change

Please check the option that is related to your PR.

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
  - In this case, we recommend to discuss your modification on GitHub issues before creating the PR
- [x] Documentations (modification for documents)

## Checklist

- [x] My code follows [the style guidelines of this project](https://deepchem.readthedocs.io/en/latest/development_guide/coding.html)
  - [x] Run `yapf -i <modified file>` and check no errors (**yapf version must be  0.32.0**)
  - [x] Run `mypy -p deepchem` and check no errors
  - [x] Run `flake8 <modified file> --count` and check no errors
  - [x] Run `python -m doctest <modified file>` and check no errors
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New unit tests pass locally with my changes
- [x] I have checked my code and corrected any misspellings

### CODE REVIEW

**Review by rbharath — COMMENTED** [review]
@aaronrockmenezes Excited to see this! A quick reminder to make sure to add some additional unit tests to check this new functionality

**Comment by aaronrockmenezes** [comment]
@rbharath I had a doubt. The original code used early stopping by default. Wouldn't it be better if we kept it as an option? It would make the code a little longer, but seems fair. 
Also, could you please help me out with what unit tests I should include? (I haven't ever done unit testing :/ )

**Comment by aaronrockmenezes** [comment]
@rbharath Added test to keep early stopping greater than 1. 
I thought of adding a test to check if early stopping is triggered but can't do that as it need not always trigger within the given number of estimators/epochs. 
In any case, if we think a test a necessary, I have code to check if the best iteration comes before the total number of iterations.

File: deepchem/models/gbdt_models/gbdt_model.py [hunk 1]
```diff
@@ -55,6 +58,16 @@ def __init__(self,
         self.early_stopping_rounds = early_stopping_rounds
         self.model_type = self._check_model_type()
 
+        if self.early_stopping_rounds<=0:
+            raise ValueError("Early Stopping Rounds cannot be less than 1.")
+
+        if self.model.__class__.__name__.startswith('XGB'):
```
Comment (rbharath) [inline deepchem/models/gbdt_models/gbdt_model.py:64]: Let's add a unit test for callbacks

File: deepchem/models/gbdt_models/gbdt_model.py [hunk 2]
```diff
@@ -107,16 +120,19 @@ def fit(self, dataset: Dataset):
...
+        self.model.fit(
+            X_train,
+            y_train,
+            callbacks=[self.callbacks],
+            eval_metric=self.eval_metric,
+            eval_set=[(X_test, y_test)])
 
         # retrain model to whole data using best n_estimators * 1.25
         if self.model.__class__.__name__.startswith('XGB'):
+            assert self.model.best_iteration<self.model.n_estimators
```
Comment (aaronrockmenezes) [inline deepchem/models/gbdt_models/gbdt_model.py:132]: @rbharath Added a test to check if the best iteration is less than the number of estimators. If early stopping is triggered, this condition would be true.

**Review by rbharath — COMMENTED** [review]
A few changes needed:

- Fix mypy issues
- Add explicit unit tests
- Remove assert statements and instead raise exceptions instead

**Review by rbharath — COMMENTED** [review]
Good to merge in once CI is clear

**Review by rbharath — APPROVED** [review]
LGTM

**Pull request merged**